File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/RecommendedTitlesService.kt
package com.rork.guidestreamtvandroid.data.remote

import com.rork.guidestreamtvandroid.SupabaseConfig
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * One recommended title from the `recommend_titles` edge function.
 * Mirrors iOS `RecommendedTitle`.
 */
@Serializable
data class RecommendedTitle(
    @SerialName("tmdb_id") val tmdbId: Int = 0,
    @SerialName("media_type") val mediaType: String = "tv",
    val title: String = "",
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("vote_average") val voteAverage: Double? = null,
    @SerialName("match_percentage") val matchPercentage: Int = 0,
) {
    val isTv: Boolean get() = mediaType == "tv"

    /** Full TMDB poster URL at the width the other home rails request. */
    val posterUrl: String?
        get() = posterPath?.takeIf { it.isNotBlank() }?.let { "https://image.tmdb.org/t/p/w500$it" }
}

@Serializable
data class RecommendedTitlesResponse(
    val items: List<RecommendedTitle> = emptyList(),
    val cached: Boolean = false,
    @SerialName("seed_count") val seedCount: Int = 0,
    val reason: String? = null,
)

/**
 * Client for the `recommend_titles` edge function (deployed with
 * verify_jwt=false), which backs the "Recommended for You" home rail.
 * Companion to [RecommendedCreatorsService] and built the same way.
 *
 * The server reads the signals itself — watchlist, likes, watched, release
 * reminders, trailer engagement, browse events, favourite teams, followed
 * creators — so this client sends only identity and the subscribed-service set.
 * It never assembles a taste profile, which is what keeps iOS, tvOS and Android
 * ranking identically.
 *
 * Freshness is the server's job too: it caches per owner against a fingerprint
 * of that owner's signals, so calling this on every Home load is cheap. An
 * unchanged signal set returns the cached rail with no TMDB calls; a title saved
 * a moment ago rebuilds it on the very next call.
 */
object RecommendedTitlesService {

    /** The home rail's size. */
    const val RAIL_LIMIT = 20

    /**
     * @param userId Supabase auth uuid, or null for a guest.
     * @param deviceId Device identifier, used as the owner when signed out.
     * @param subscribedServices The viewer's saved services. The server filters
     *   every recommendation to these, so an empty set legitimately returns an
     *   empty rail rather than titles the viewer cannot watch.
     * @return the ranked titles, or an empty list on any failure — the rail is
     *   additive and must never be able to break Home.
     */
    suspend fun recommend(
        userId: String?,
        deviceId: String,
        subscribedServices: List<String>,
        limit: Int = RAIL_LIMIT,
    ): List<RecommendedTitle> {
        if (subscribedServices.isEmpty()) return emptyList()
        return try {
            val client = HttpClient {
                install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
            }
            val url = "${SupabaseConfig.URL.trim()}/functions/v1/recommend_titles"
            val body = buildJsonObject {
                if (!userId.isNullOrBlank()) {
                    put("userId", JsonPrimitive(userId))
                } else {
                    put("deviceId", JsonPrimitive(deviceId))
                }
                put("subscribedServices", JsonArray(subscribedServices.map { JsonPrimitive(it) }))
                put("limit", JsonPrimitive(limit))
            }
            val response: HttpResponse = client.post(url) {
                contentType(ContentType.Application.Json)
                header(HttpHeaders.ContentType, "application/json")
                header("apikey", SupabaseConfig.ANON_KEY)
                header(HttpHeaders.Authorization, "Bearer ${SupabaseConfig.ANON_KEY}")
                setBody(body.toString())
            }
            if (response.status.value == 200) {
                response.body<RecommendedTitlesResponse>().items
            } else {
                emptyList()
            }
        } catch (_: Exception) {
            emptyList()
        }
    }
}

File: ios/GuideStreamTV/Services/RecommendedTitlesService.swift
//
//  RecommendedTitlesService.swift
//  GuideStreamTV
//
//  Client for the `recommend_titles` Supabase edge function, which backs the
//  "Recommended for You" home rail. Companion to RecommendedCreatorsService —
//  same shape, same reasoning: the scorer lives on the server so iOS, tvOS and
//  Android rank identically and the algorithm cannot drift between platforms.
//
//  The server reads the signals directly (watchlist, likes, watched, release
//  reminders, trailer engagement, browse events, favourite teams, followed
//  creators), so this client sends only identity and the subscribed-service
//  set — it never assembles a taste profile itself.
//
//  Freshness is the server's job too. It keys a cache on a fingerprint of the
//  owner's signals, so calling this on every Home appearance is cheap: an
//  unchanged signal set returns the cached rail with no TMDB calls, and saving
//  a title or liking a trailer rebuilds it on the very next call. That is what
//  makes the rail "update when the signals change" without any client polling.
//

import Foundation

/// One recommended title as returned by `recommend_titles`.
nonisolated struct RecommendedTitle: Decodable, Sendable, Identifiable {
    let tmdbId: Int
    let mediaType: String
    let title: String
    let posterPath: String?
    let backdropPath: String?
    let voteAverage: Double?
    let matchPercentage: Int

    var id: String { "\(mediaType):\(tmdbId)" }
    var isTV: Bool { mediaType == "tv" }

    /// Full TMDB poster URL at the same width the other home rails request.
    var posterUrl: String? {
        guard let posterPath, !posterPath.isEmpty else { return nil }
        return "https://image.tmdb.org/t/p/w500\(posterPath)"
    }

    enum CodingKeys: String, CodingKey {
        case tmdbId = "tmdb_id"
        case mediaType = "media_type"
        case title
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case voteAverage = "vote_average"
        case matchPercentage = "match_percentage"
    }
}

nonisolated enum RecommendedTitlesService {

    private struct Response: Decodable {
        let items: [RecommendedTitle]
        let cached: Bool?
        let seedCount: Int?
        let reason: String?

        enum CodingKeys: String, CodingKey {
            case items, cached, reason
            case seedCount = "seed_count"
        }
    }

    /// Fetches the recommended rail for the current viewer.
    ///
    /// - Parameters:
    ///   - userId: Supabase auth uuid, or nil for a guest.
    ///   - deviceId: Device identifier, used as the owner when signed out.
    ///   - subscribedServices: The viewer's saved services. The server filters
    ///     every recommendation to these, so an empty set legitimately returns
    ///     an empty rail rather than titles the viewer cannot watch.
    ///   - limit: Maximum rail length.
    /// - Returns: The ranked titles, `[]` when the server genuinely has no
    ///   recommendations, and **`nil` when the call failed** — a non-200, a
    ///   decode error, or a cancelled request.
    ///
    ///   The two used to be the same value, and the caller assigned it either
    ///   way. A pull-to-refresh that got cancelled mid-flight therefore wiped
    ///   a rail that was already on screen (GUI-98 follow-up). Failure has to be
    ///   distinguishable from emptiness for the caller to keep what it has.
    static func fetch(
        userId: String?,
        deviceId: String,
        subscribedServices: [String],
        limit: Int = 20
    ) async -> [RecommendedTitle]? {
        guard !subscribedServices.isEmpty else { return [] }

        let base = SupabaseConfig.url.trimmingCharacters(in: .whitespaces)
        guard let url = URL(string: "\(base)/functions/v1/recommend_titles") else { return nil }

        var body: [String: Any] = [
            "subscribedServices": subscribedServices,
            "limit": limit
        ]
        if let userId, !userId.isEmpty {
            body["userId"] = userId
        } else {
            body["deviceId"] = deviceId
        }
        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(SupabaseConfig.anonKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = httpBody

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            let decoded = try JSONDecoder().decode(Response.self, from: data)
            return decoded.items
        } catch {
            return nil
        }
    }
}
